Corrects ADU upstream air leakage model to use the maximum available flow rate for CV air terminals

Nigusse opened the pull request:
Pull request overview
---------------------
 - Fixes #10110 
 - Corrects the upstream air leakage rate calculation at the Air Distibution Unit (ADU). Modified the upstream air leakage rate calculation on the basis of the design flow rate for the VAV air terminal and the maximum available mass flow rate for the CV air terminal. Currently, the air leakage flow is a fraction of the design flow rate and is constant. This fix makes the air leakage amount calculation dependent on the air terminal type. This change impacts the CV air terminal types only.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Nigusse committed:
Initial Fix

Nigusse committed:
Cleanup and removed unused variables

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

Nigusse commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -2088,19 +2088,8 @@ void InitAirLoops(EnergyPlusData &state, bool const FirstHVACIteration) // TRUE
-                    state.dataAirLoop->AirLoopFlow(AirLoopNum).DesSupply / state.dataSimAirServingZones->SumZoneDesFlow;
-            } else {
-                state.dataAirLoop->AirLoopFlow(AirLoopNum).SysToZoneDesFlowRatio = 1.0;
-            }
```
The variable `SysToZoneDesFlowRatio` is no longer needed as the upstream air leakage rate is now calculated directly from the maximum available air flow rate. This variable was used for upstream ADU air leakage rate calculation only.

Nigusse commented on src/EnergyPlus/SimAirServingZones.hh:
```diff
@@ -237,7 +237,6 @@ struct SimAirServingZonesData : BaseGlobalStruct
 
     int TUInNode = 0;            // inlet node number of a terminal unit
-    Real64 SumZoneDesFlow = 0.0; // sum of the zone design air mass flow rates for zones served by a system
     Real64 OAReliefDiff = 0.0;   // local for massflow change across OA system, kg/s
```
Removed local variable, which was used to calculate `SysToZoneDesFlowRatio`, no longer needed.

Nigusse commented on src/EnergyPlus/SimAirServingZones.hh:
```diff
@@ -305,7 +304,6 @@ struct SimAirServingZonesData : BaseGlobalStruct
         this->Vot = 0.0;
 
         this->TUInNode = 0;         // inlet node number of a terminal unit
-        this->SumZoneDesFlow = 0.0; // sum of the zone design air mass flow rates for zones served by a system
```
Same as above.

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -690,14 +690,8 @@ namespace ZoneAirLoopEquipmentManager {
-                        DesFlowRatio = state.dataAirLoop->AirLoopFlow(AirLoopNum).SysToZoneDesFlowRatio;
-                    } else {
-                        DesFlowRatio = 1.0;
-                    }
```
Not needed any longer.

Nigusse replied:
Depending on the value of the `IsConstLeakageRate` flag either a constant leakage flow rate or a proportional leakage flow rate will be calculated. Four `VAV` air terminals will use a constant leakage flow rate (assuming there is duct static pressure reset), and two `CV` air terminals will use a proportional leakage flow rate. Currently, only `six` air terminals support the air leakage model.

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -690,14 +690,8 @@ namespace ZoneAirLoopEquipmentManager {
                     MassFlowRateUpStreamLeakMax = max(state.dataDefineEquipment->AirDistUnit(AirDistUnitNum).UpStreamLeakFrac *
-                                                          state.dataLoopNodes->Node(InNodeNum).MassFlowRateMax * DesFlowRatio,
+                                                          state.dataLoopNodes->Node(InNodeNum).MassFlowRateMaxAvail,
                                                       0.0);
```
The upstream leakage rate is directly calculated from the maximum available air flow rate. No need for local variable `DesFlowRatio`.

rraustad replied:
We did talk about this and I agreed with you at the time. But as I think about this more the leakage in a duct would be driven by duct static pressure and E+ assumes that the duct static pressure remains constant. This method assumes duct static pressure varies with air flow rate. If I blocked off one zone diffuser in a VAV system would there be leakage in that section of duct (develop says yes) or not (this method says no)?

rraustad replied:
It may also be that a VAV system's leakage response is what develop does now. And a constant volume system using a cycling fan would respond as this branch suggests.

Nigusse replied:
You brought up interesting points: 
- If a static pressure rest is considered, the leakage amount is constant, not the leakage fraction, because the leakage amount is driven by the pressure difference, which is constant with static reset.
- otherwise, the leakage amount may vary with the duct pressure, which is dependent on the supply air flow rate

Hmmm, the documentation is also consistent with the current code.

Nigusse commented on src/EnergyPlus/DataAirLoop.hh:
```diff
@@ -163,7 +163,6 @@ namespace DataAirLoop {
         // Members
         Real64 DesSupply = 0.0;             // design supply air mass flow rate for loop [kg/s]
         Real64 DesReturnFrac = 1.0;         // the design return flow rate as a fraction of supply flow assuming no exhaust (0 to 1)
-        Real64 SysToZoneDesFlowRatio = 0.0; // System design flow divided by the sum of the zone design flows
```
This variable was used for upstream ADU leakage rate calculation only, with this fix it is no longer needed.

Nigusse reviewed (commented):
Code change walk-through for ADU upstream air leakage rate calculation fix:
- the ADU upstream air leakage rate is now calculated directly from the air inlet node maximum available mass flow rate

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -669,11 +669,11 @@ namespace ZoneAirLoopEquipmentManager {
         Real64 MassFlowRateMinAvail;        // min avail mass flow rate excluding leaks [kg/s]
         Real64 MassFlowRateUpStreamLeakMax; // max upstream leak flow rate [kg/s]
-        Real64 DesFlowRatio(0.0);           // ratio of system to sum of zones design flow rate
+        //Real64 DesFlowRatio(0.0);           // ratio of system to sum of zones design flow rate
```
Forgot to remove the commented-out variables, will be removed.

Nigusse committed:
Clang format and cleanup of unused variables

Nigusse committed:
More code cleanup

Nigusse commented:
A VAV system serving single zone with VAV Air Terminal was created for illustration. Below are plots of supply air flow rate entering and leaving the ADU. A 5% and 7% leakage rates were applied to the upstream and downstream of the ADU, respectively. `Node 2` is the ADU inlet node and `Node 5` is the ADU air outlet node. The leakage rate for the branch is lower than that of develop, because the leakage rates are calculated based on maximum available mass flow rate of the inlet node, instead of the maximum mass flow rate. 

![ADU_inlet_outlet_FlowRate_wLeak](https://github.com/NREL/EnergyPlus/assets/5805126/b849223b-ce54-4129-9d65-a10a305a9584)

Nigusse commented:
The two plots below are zone air temperature and the zone air sensible heating rates for the same defect file. The zone air temperature is maintained at the thermostat heating setpoints in both cases (develop and branch) but the branch delivering higher heating rate due to reduced leakage rate consistent with the specified leakage fractions.


![ZoneAirTemp_wLeak](https://github.com/NREL/EnergyPlus/assets/5805126/dace97d1-6afa-4bbd-98ba-bf054e8f4524)

Nigusse commented:
Similar trend was observed for a single zone served with a VAV system with CV reheat air terminal defect file. The first plot shows higher leakage rate for the develop compared to the branch for the same user specified leakage fractions.

![ADU_inlet_outlet_FlowRate_CAV_wLeak](https://github.com/NREL/EnergyPlus/assets/5805126/7b022543-077c-4f69-bc74-22b861031039)


In the case of CV reheat air terminal, both develop and branch could not provide enough heating rate to maintain the thermostat heating setpoint but develop performed worst due to higher air leakage rate as shown below.

![ZoneAirTemp_CAVwLeak](https://github.com/NREL/EnergyPlus/assets/5805126/97451d24-1fb9-466a-adc6-58b625699259)

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

nrel-bot-2b commented:
@Nigusse it has been 12 days since this pull request was last updated.

nrel-bot-2 commented:
@Nigusse it has been 7 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into Issue10110_ADU_UpstreamLeakage_calc

mjwitte commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -2088,19 +2088,8 @@ void InitAirLoops(EnergyPlusData &state, bool const FirstHVACIteration) // TRUE
-            state.dataSimAirServingZones->SumZoneDesFlow = 0.0;
             state.dataAirLoop->AirLoopFlow(AirLoopNum).DesSupply = thisPrimaryAirSys.DesignVolFlowRate * state.dataEnvrn->StdRhoAir;
             state.dataAirLoop->AirLoopFlow(AirLoopNum).DesReturnFrac = thisPrimaryAirSys.DesignReturnFlowFraction;
-            for (int ZoneInSysIndex = 1; ZoneInSysIndex <= thisAirToZoneNodeInfo.NumZonesCooled; ++ZoneInSysIndex) {
```
Always nice to delete unnecessary code!

mjwitte reviewed (commented):
@Nigusse This looks good. Did you say this needs some documentation? I'm merging in develop to get some fresh CI results.

mjwitte committed:
format

mjwitte committed:
remove unused

Nigusse committed:
Merged in develop; resolved conflict

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

Nigusse commented:
@mjwitte There is an issue brought by @rraustad that needs further discussion or consideration. The current code assumes that the leakage rate is constant (based on ADU design maximum flow), which is also based on static pressure reset control. I modified the code to make the leakage rate proportional to the maximum available flow. It needs discussion on which approach to use. 

One approach would be to implement both options depending on the air terminal type: (1) constant leakage amount for VAV air terminals, which assumes a static pressure reset is present, or (2) leakage rate proportional to the maximum available flow for other air terminal types that support air leakage model.

Nigusse commented:
Below is the list of air terminals that support ADU's air leakages model. 

```
ADU Leakage Model Supported In:
 I: Fixed leakage amount (assumes static pressure reset available, current implementation)
  (1) AirTerminal:SingleDuct:VAV:Reheat
  (2) AirTerminal:SingleDuct:VAV:NoReheat
  (3) AirTerminal:SingleDuct:VAV:HeatAndCool:Reheat
  (4) AirTerminal:SingleDuct:VAV:HeatAndCool:NoReheat
II: Proportional leakage amount (assumes static pressure reset not available)
  (5) AirTerminal:SingleDuct:ConstantVolume:Reheat
  (6) AirTerminal:SingleDuct:ConstantVolume:NoReheat
 
```

Nigusse committed:
Added flag for calculating Leakage Flow Rate

Nigusse committed:
Clang format fix

Nigusse committed:
Corrected the logic of new flag var use

Nigusse committed:
Merge branch 'Issue10110_ADU_UpstreamLeakage_calc' of https://github.com/NREL/EnergyPlus into Issue10110_ADU_UpstreamLeakage_calc

Nigusse committed:
Reverted changes for VAV air terminals

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

Nigusse commented:
Rolled back the changes for VAV air terminals. Now, the fix applies to CV air terminals only. Summary of changes: 
- (1) ADU with VAV air terminals use current leakage calculation, i.e., upstream leakage amount is constant based on design maximum flow rate
- (2) ADU with CV air terminals use proportional to the maximum available flow
- (3) Downstream leakage amount is proportional to the supply air flow rate for VAV and CV air terminals (not changed)

This change impacts air leakage model of ADU with CV air terminals only. @mjwitte @Myoldmopar @rraustad any concern about this approach?

Nigusse commented:
The following three test Idfs, which have CV air terminal units, are expected to show diffs. 
`5ZoneCAV_MaxTemp.idf
5ZoneCAV_MaxTemp_JSON_Outputs.idf
5ZoneCAVtoVAVWarmestTempFlow.idf`

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

Nigusse committed:
Updated IO Ref doc; added missing air terminals supported in ADU

Nigusse committed:
Updated Air Distribution Units IO Ref document

Nigusse commented on doc/input-output-reference/src/overview/group-zone-equipment.tex:
```diff
@@ -57,7 +57,9 @@ \subsection{ZoneHVAC:AirDistributionUnit}\label{zonehvacairdistributionunit}
-  \hyperref[airterminalsingleductconstantvolumereheat]{AirTerminal:SingleDuct:ConstantVolume:Reheat}
+  \hyperref[airterminalsingleductconstantvolumereheat]{AirTerminal:SingleDuct:ConstantVolume:Reheat}  
+\item
+  \hyperref[airterminalsingleductconstantvolumenoreheat]{AirTerminal:SingleDuct:ConstantVolume:NoReheat}
```
Added missing `AirTerminal:SingleDuct:ConstantVolume:NoReheat` air terminal options of ADU.

Nigusse commented on doc/input-output-reference/src/overview/group-zone-equipment.tex:
```diff
@@ -68,6 +70,8 @@ \subsection{ZoneHVAC:AirDistributionUnit}\label{zonehvacairdistributionunit}
 \item
   \hyperref[airterminalsingleductconstantvolumefourpipeinduction]{AirTerminal:SingleDuct:ConstantVolume:FourPipeInduction}
+\item
+  \hyperref[airterminalsingleductconstantvolumefourpipebeam]{AirTerminal:SingleDuct:ConstantVolume:FourPipeBeam}
```
Added missing `AirTerminal:SingleDuct:ConstantVolume:FourPipeBeam` air terminal options of ADU.

Nigusse commented on doc/input-output-reference/src/overview/group-zone-equipment.tex:
```diff
@@ -85,7 +89,25 @@ \subsection{ZoneHVAC:AirDistributionUnit}\label{zonehvacairdistributionunit}
 
-The Air Distribution unit also allows the user to specify leaks in the supply air duct system. These inputs are used in the EnergyPlus Simplified Duct Leakage Model (SDLM). This model simulates a specific configuration: supply leaks to a return plenum in a commercial VAV or CV system. The system must have a constant static pressure setpoint. Within these limitations SDLM allows the user to easily ...
+The Air Distribution unit also allows the user to specify leaks in the supply air duct system. These inputs are used in the EnergyPlus Simplified Duct Leakage Model (SDLM). This model simulates a specific configuration: supply leaks to a return plenum in a commercial VAV or CV system. The VAV system must have a constant static pressure setpoint, and a constant upstream air leakage flow rate is ca ...
+
```
Updated the document to describe how upstream leakage rates are calculated depending on the air terminal type.

Nigusse commented on doc/input-output-reference/src/overview/group-zone-equipment.tex:
```diff
@@ -85,7 +89,25 @@ \subsection{ZoneHVAC:AirDistributionUnit}\label{zonehvacairdistributionunit}
+\item
+  \hyperref[airterminalsingleductconstantvolumenoreheat]{AirTerminal:SingleDuct:ConstantVolume:NoReheat}
+
+\end{itemize}
```
Added list of air terminals currently supported in the simplified air leakage model.

Nigusse commented on doc/input-output-reference/src/overview/group-zone-equipment.tex:
```diff
@@ -107,7 +129,7 @@ \subsubsection{Inputs}\label{inputs-055}
 
-This is the leakage upstream of the terminal unit as a fraction of the design flow rate through the unit. It is the leakage fraction at the design flow rate. It is used to calculate a leakage flow rate which is then held constant while the system air flow varies. This input is optional; the default is zero.
+This is the leakage upstream of the terminal unit as a fraction of the design flow rate or maximum available flow rate through the unit. It is the leakage fraction at the design flow rate for the VAV air terminal and at the current maximum available flow rate for the CV air terminals. For the VAV air terminal, it is used to calculate a leakage flow rate, which is then held constant while the syst ...
 
```
Updated upstream leakage rate calculation method depending on the air terminal type.

Nigusse reviewed (commented):
Updated I/O reference document of ADU regarding the upstream leakage rate calculation.

Nigusse commented:
@mjwitte I modified the code to account for @rraustad comment to separate the leakage rate calculation and updated the IO ref documentation accordingly. This is ready for review.

rraustad commented:
> Rolled back the changes for VAV air terminals. Now, the fix applies to CV air terminals only. Summary of changes:
> 
> * (1) ADU with VAV air terminals use current leakage calculation, i.e., upstream leakage amount is constant based on design maximum flow rate
> * (2) ADU with CV air terminals use proportional to the maximum available flow
> * (3) Downstream leakage amount is proportional to the supply air flow rate for VAV and CV air terminals (not changed)
> 
> This change impacts air leakage model of ADU with CV air terminals only. @mjwitte @Myoldmopar @rraustad any concern about this approach?

I had the opportunity to think about the fan curves and associated system curve. I think (1) above covers VAV correctly. And (2) covers single-speed with cycling fans. (3) appears to cover downsteam leakage from the ADU. I think what is missing, and may be addressed in a different branch if desired, is multi-speed systems that have different max air flow rates for each speed. The figure below shows what I believe is the static-pressure curve with air flow. Red is (1), purple is (2). The blue (including purple) is what I think should happen with multi-speed systems that have a fan with some maximum air flow rate that is not indicative of the system's air flow rate at each speed where static pressure would change. If leakage is proportional to static pressure then we need to know what the static pressure is at each speed, which is a squared fan curve law and could be calculated without the need for additional user input.

![image](https://github.com/NREL/EnergyPlus/assets/5778156/c6550796-f493-42fe-b9ba-0bdd9cd7fe47)

Nigusse committed:
Applied ShortCuts

Nigusse committed:
Added string_view array for enum class ZnAirLoopEquipType

Nigusse commented:
One additional issue raised by @RR ADU air leakage leakage for multi-speed system. I will post a new issue for calculating air leakage at ADU with a multi-speed system.

Nigusse committed:
Merge branch 'develop' into Issue10110_ADU_UpstreamLeakage_calc

rraustad commented:
@Nigusse also note in the figure that the VAV TU air flow rate could go to 0 (when Minimum Air Flow Fraction = 0) while the fan is still on, which would mean there would still be upstream duct leakage. The only time the leakage = 0 is when the fan is off.

Nigusse committed:
string_view array def cleanup

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -85,6 +85,44 @@ namespace ZoneAirLoopEquipmentManager {
+        "AIRTERMINAL:SINGLEDUCT:USERDEFINED",
+        "AIRTERMINAL:SINGLEDUCT:MIXER",
+        "AIRTERMINAL:SINGLEDUCT:CONSTANTVOLUME:FOURPIPEBEAM"};
+
```
Adds air terminal equipment name string_view array definition for use with Enum Class ZnAirLoopEquipType.

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -201,6 +239,7 @@ namespace ZoneAirLoopEquipmentManager {
         if (NumAirDistUnits > 0) {
 
             for (AirDistUnitNum = 1; AirDistUnitNum <= NumAirDistUnits; ++AirDistUnitNum) {
+                auto &airDistUnit = state.dataDefineEquipment->AirDistUnit(AirDistUnitNum);
```
applied ShortCut

Nigusse commented on src/EnergyPlus/ZoneAirLoopEquipmentManager.cc:
```diff
@@ -215,341 +254,191 @@ namespace ZoneAirLoopEquipmentManager {
+                case DataDefineEquip::ZnAirLoopEquipType::SingleDuctATMixer:
+                    if (airDistUnit.UpStreamLeak || airDistUnit.DownStreamLeak) {
+                        ShowSevereError(state, format("Error found in {} = {}", CurrentModuleObject, airDistUnit.Name));
                         ShowContinueError(state,
```
Replaced `IF logic` with `switch` using the enum class, air terminal string_view array, and getEnumValue() function.

Nigusse reviewed (commented):
Applied ShortCuts and added string_view array for air terminal equipment.
- Cleanup get inputs, and other functions

Nigusse committed:
Merge branch 'Issue10110_ADU_UpstreamLeakage_calc' of https://github.com/NREL/EnergyPlus into Issue10110_ADU_UpstreamLeakage_calc

Nigusse committed:
Added unit test

Nigusse committed:
Remove enumeration value warning error message

Nigusse committed:
Removed unused variable

Nigusse committed:
Fixed the new unit test failed on Ubuntu

Nigusse committed:
More ShortCuts and Cleanup

Nigusse commented:
@Myoldmopar Any clue why the Ubuntu build is failing?

Myoldmopar commented:
From a quick look, my guess is that you are trying to do an EXPECTEQ or some other assertion on an enum in a unit test. Make sure you use compare_enums if so. I'll check it out shortly if you still have issues.

Nigusse committed:
Fixed enum comparison of a unit test

Nigusse commented:
Yes, I am doing enum compare in the unit test and will use compare_enums instead @Myoldmopar Thanks.

Myoldmopar commented:
Windows build failure is almost certainly unrelated, I don't even think we need to delete the results to get that refreshed as far as CI results go.  This branch does have a few diffs, so I'm going to hold off from doing much here to avoid polluting the other CI results.

Nigusse commented:
Okay. There are three example files with diffs, and the diffs are expected. These diffs are caused by the change in the leakage rate calculation method for ADU with CV air terminal unit.

Myoldmopar commented:
I'll pull develop in and do a quick test.  The diffs have been discussed, evidence shown as to why it is better, and no need to hold this up.

Myoldmopar commented:
Yep, all good, merging.  Thanks @Nigusse
